### ISSUE TITLE
pre-meteor.js will raise TypeError in the face of inbound requests without Host headers

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#1215]
Jan\ on IRC reports this issue:

```
TypeError: Cannot call method 'split' of undefined
    at dispatchToMeteorOrStaticPublishing (app/server/pre-meteor.js:424:37)
    at Server.redirectToMeteorOrServeStaticPublishing (app/server/pre-meteor.js:414:12)
    at Server.emit (events.js:98:17)
    at HTTPParser.parser.onIncoming (http.js:2113:12)
    at HTTPParser.parserOnHeadersComplete [as onHeadersComplete] (http.js:122:23)
    at Socket.socket.ondata (http.js:1971:22)
    at TCP.onread (net.js:528:27)

    app/server/pre-meteor.js:424
        var hostname = req.headers.host.split(":")[0];
```

@zarvox rightly notices that very few clients will not send a Host header. Also Sandstorm depends on a valid Host header for proper functioning generally.

Regardless I don't like it when code I wrote crashes! So I'm filing this.

One thing I would love to see result from this bug is someone finding a tool that can automatically find this error. (I guess it would also complain about `req.headers` being possibly undefined.) If this error can be found by some static code checker, then I'd be very +1 on fixing it as part of fixing issues that the static code checker found.

**Comment by Jan02** [comment]
thank you for filing this for me.

**Comment by kentonv** [comment]
FWIW, I believe `Host:` is actually required by HTTP/1.1, so feel free to throw a "400 Bad Request" if it's missing.

I'd be surprised if a static code checker can find this. How is it supposed to know that, e.g., `req.headers` is always defined but `req.headers.host` isn't necessarily? Seems like you need a type system for this. I wish we had a type system. :/

**Comment by dwrensha** [comment]
Fixed by #1471.
